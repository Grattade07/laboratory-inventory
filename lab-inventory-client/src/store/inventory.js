import {createSlice, current} from "@reduxjs/toolkit"

/* state slice for the application */
export const inventorySlice = createSlice({
    name: "inventory",

    /* initial set states */
    initialState: {
        inventoryUpdateCount: 0,
        inventoryList: [],
        displayedInventoryList: [],
        userJWToken: "",
        isAdmin: false,
        isLoggedIn: false
    },

    /* reducers to change state */
    reducers: {
        /* increments the update count */
        updateInventoryCount: (state) => {
            state.inventoryUpdateCount += 1
        },
        /* updates the inventory list */
        updateInventoryList: (state, action) => {
            state.inventoryList = action.payload
        },
        /* updates to the default inventory list to be shown in the table */
        displayList: (state) => {
            state.displayedInventoryList = state.inventoryList
        },
        /* makes the displayed table sorted by item name */
        displayItemSortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.name > p2.name) ? 1 : (p1.name < p2.name) ? -1 : 0)
        },
        /* makes the displayed table sorted by item location */
        displayLocationSortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.location > p2.location) ? 1 : (p1.location < p2.location) ? -1 : 0)
        },
        /* makes the displayed table sorted by item batch */
        displayBatchSortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.batch > p2.batch) ? 1 : (p1.batch < p2.batch) ? -1 : 0)
        },
        /* makes the displayed table sorted by item quantity */
        displayQuantitySortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.quantity > p2.quantity) ? 1 : (p1.quantity < p2.quantity) ? -1 : 0)
        },
        /* makes the displayed table sorted by item expiry */
        displayExpirySortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.expiryDate > p2.expiryDate) ? 1 : (p1.expiryDate < p2.expiryDate) ? -1 : 0)
        },
        /* updates the stored JWToken */
        addJWToken: (state, action) => {
            state.userJWToken = action.payload
        },
        /* removes the stored JWToken */
        removeJWToken: (state) => {
            state.userJWToken = ""
        },
        /* changes the displayed table to what was search in the search bar */
        displayFilteredList: (state, action) => {
            let filterString = action.payload
            /* current function gets the current state of the inventoryList state. Found this function from url: https://redux-toolkit.js.org/api/other-exports#current */
            let inventoryList = current(state.inventoryList)
            let filteredList = function(string) {
                let filteredList = []

                for (let i = 0; i < inventoryList.length; i++) {
                    if(inventoryList[i].name.toLowerCase().includes(string)) {
                        filteredList.push(inventoryList[i])
                    }
                }

                return filteredList
            }

            state.displayedInventoryList = filteredList(filterString)
        },
        /* resets the displayed list to the original data */
        resetDisplayList: (state) => {
            state.displayedInventoryList = state.inventoryList
        },
        /* updates if user is admin */
        setIsAdmin: (state, action) => {
            state.isAdmin = action.payload
        },
        /* updates if user is logged in */
        setIsLoggedIn: (state, action) => {
            state.isLoggedIn = action.payload
        }
    }
})

export const {
    updateInventoryCount, 
    updateInventoryList, 
    displayList, 
    displayItemSortedList, 
    displayLocationSortedList, 
    displayBatchSortedList, 
    displayQuantitySortedList, 
    displayExpirySortedList, 
    addJWToken, 
    removeJWToken, 
    displayFilteredList, 
    resetDisplayList,
    setIsAdmin,
    setIsLoggedIn
} = inventorySlice.actions

export default inventorySlice.reducer