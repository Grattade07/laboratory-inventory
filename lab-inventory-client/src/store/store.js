import { configureStore } from "@reduxjs/toolkit";
import inventoryReducer from "./inventory"

/* store for application */
export default configureStore({
    reducer: {
        inventory : inventoryReducer
    }
})