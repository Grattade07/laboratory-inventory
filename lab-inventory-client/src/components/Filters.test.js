import renderer from "react-test-renderer"
import Filter from "./Filters"
import { Provider } from 'react-redux';
import store from '../store/store';

/* tests that the filter component renders correctly  */
it("renders correctly", () => {
    const tree = renderer
        .create(<Provider store={store}>
                    <Filter />
                </Provider>
        )
        .toJSON()
    expect(tree).toMatchSnapshot()
})