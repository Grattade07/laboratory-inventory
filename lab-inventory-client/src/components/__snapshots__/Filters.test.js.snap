// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  id="filters-container"
>
  <h4
    className="underline"
  >
    Sort By:
  </h4>
  <form
    className=""
    id="filter-list"
  >
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        defaultChecked={true}
        disabled={false}
        name="filter"
        onClick={[Function]}
        type="radio"
      />
      <label
        className="form-check-label"
        title=""
      >
        None
      </label>
    </div>
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        disabled={false}
        name="filter"
        onClick={[Function]}
        type="radio"
      />
      <label
        className="form-check-label"
        title=""
      >
        Item
      </label>
    </div>
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        disabled={false}
        name="filter"
        onClick={[Function]}
        type="radio"
      />
      <label
        className="form-check-label"
        title=""
      >
        Location
      </label>
    </div>
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        disabled={false}
        name="filter"
        onClick={[Function]}
        type="radio"
      />
      <label
        className="form-check-label"
        title=""
      >
        Batch
      </label>
    </div>
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        disabled={false}
        name="filter"
        onClick={[Function]}
        type="radio"
      />
      <label
        className="form-check-label"
        title=""
      >
        Quantity
      </label>
    </div>
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        disabled={false}
        name="filter"
        onClick={[Function]}
        type="radio"
      />
      <label
        className="form-check-label"
        title=""
      >
        Expiry Date
      </label>
    </div>
  </form>
</section>
`;
